units/event_log_tests.cc

```diff
@@ -0,0 +1,132 @@
+#include <array>
+#include <cassert>
+#include <cstddef>
+#include <cstdint>
+#include <cstdio>
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include "gtest/gtest.h"
+
+// TODO(miladhakimi): Need better naming.
+enum LogLevel { CommonFile, PrivateFile };
+enum ValueType { String, Int64, Vector };
+
+std::string ValueToString(const std::string &value) { return value; }
+
+std::string ValueToString(const int64_t &value) {
+  return std::to_string(value);
+}
+
+// TODO(miladhakimi): This can be different for other formats.
+template <typename T>
+std::string ValueToString(const T &values) {
+  std::string tmp = "\"[";
+  using std::begin;
+  using std::end;
+  auto e = end(values);
+  for (auto it = begin(values); it != e; ++it) {
+    const char *delimiter = std::next(it) != values.end() ? "," : "";
+    tmp += std::to_string(*it) + delimiter;
+  }
+  tmp += "]\"";
+  return tmp;
+}
+
+struct Attribute {
+  const char *name;
+  ValueType value_type;
+
+  template <typename T>
+  T *cast() {
+    return static_cast<T *>(this);
+  };
+};
+
+template <typename T, ValueType VT>
+struct AttributeImpl : Attribute {
+  T value;
+
+  AttributeImpl(const char *_name, T _value)
+      : Attribute{_name, VT}, value(_value) {}
+  T &GetValue() { return value; }
+};
+
+using VectorAttr = AttributeImpl<std::vector<int64_t>, ValueType::Vector>;
+using StringAttr = AttributeImpl<std::string, ValueType::String>;
+using Int64Attr = AttributeImpl<int64_t, ValueType::Int64>;
+
+struct EventLog {
+  const char *name;
+  std::vector<Attribute *> attributes;
+  LogLevel log_level;
+
+  EventLog(const char *_name, const std::vector<Attribute *> &_attributes,
+           LogLevel _log_level)
+      : name(_name), attributes(_attributes), log_level(_log_level) {}
+
+  const char *GetEventName() { return name; }
+  const std::vector<Attribute *> &GetAttributes() { return attributes; }
+  size_t GetNumAttributes() { return attributes.size(); }
+  LogLevel GetLogLevel() { return log_level; }
+};
+
+std::string EventToCSVString(EventLog &event) {
+  const std::vector<Attribute *> &attributes = event.GetAttributes();
+
+  std::string csv_str = event.GetEventName();
+  csv_str += ",";
+  size_t e;
+  for (size_t i = 0, e = attributes.size(); i != e; ++i) {
+    switch (attributes[i]->value_type) {
+      case ValueType::Int64: {
+        csv_str += ValueToString(attributes[i]->cast<Int64Attr>()->GetValue());
+        break;
+      }
+      case ValueType::String: {
+        csv_str += ValueToString(attributes[i]->cast<StringAttr>()->GetValue());
+        break;
+      }
+      case ValueType::Vector: {
+        csv_str += ValueToString(attributes[i]->cast<VectorAttr>()->GetValue());
+        break;
+      }
+      default:
```

Comment:
Some compilers warn about a default switch cases when the other cases are already exhaustive. I think we should remove this default case and instead rely on compiler warnings.